src/common/utils.ts

```diff
@@ -29,6 +30,66 @@ export const getProvider = (options: VerificationBuilderOptions): providers.Prov
   return options.provider ?? getDefaultProvider(options);
 };
 
+/**
+ * Generate Provider using the following options: (if no option is specified it will use the default values)
+ * @param {Object} ProviderDetails - Details to use for the function to successfully generate a provider.
+ * @param {string} ProviderDetails.network - The network in which the provider is connected to, i.e. "homestead", "mainnet", "ropsten", "rinkeby"
+ * @param {string} ProviderDetails.provider - Specify which provider to use: "infura", "alchemy", "etherscan" or "jsonrpc"
+ * @param {string} ProviderDetails.url - Specify which url for JsonRPC to connect to, if not specified will connect to localhost:8545
+ * @param {string} ProviderDetails.apiKey - If no apiKey is provided, a default shared API key will be used, which may result in reduced performance and throttled requests.
+ */
+export const generateProvider = (options?: ProviderDetails): providers.Provider | Signer => {
+  if (!!options && Object.keys(options).length === 1 && options.apiKey) {
+    throw new Error(
+      "We could link the apiKey provided to a provider, please state the provider to use in the parameter."
+    );
+  }
+
+  const network = options?.network || process.env.NETWORK || "homestead";
+  const provider = options?.provider || process.env.PROVIDER || "infura";
+  const url = options?.url || process.env.JSONRPC_PROVIDER_URL || "";
+
+  if (!!options && Object.keys(options).length === 1 && url) {
+    return new providers.JsonRpcProvider(url);
+  }
+
+  let apiKey = "";
+
+  switch (provider) {
+    case "infura":
+      apiKey = options?.apiKey || process.env.INFURA_API_KEY || "";
+      !apiKey &&
+        console.warn(
+          "You are using oa-verify default configuration for provider, which is not suitable for production environment. Please make sure that you configured the library correctly."
+        );
+      return apiKey ? new providers.InfuraProvider(network, apiKey) : new providers.InfuraProvider(network);
+
+    case "etherscan":
+      apiKey = options?.apiKey || process.env.ETHERSCAN_API_KEY || "";
+      !apiKey &&
+        console.warn(
+          "You are using oa-verify default configuration for provider, which is not suitable for production environment. Please make sure that you configured the library correctly."
+        );
+      return new providers.EtherscanProvider(network, apiKey);
+
+    case "alchemy":
+      apiKey = options?.apiKey || process.env.ALCHEMY_API_KEY || "";
+      !apiKey &&
```

Comment:
do the apikey check earlier instead of repeating this statement 3 times?

Reply:
ok